File: Baekjoon/Class3/34.2667 단지번호붙이기.py
from collections import deque
import sys
input = sys.stdin.readline

MOVE_DIR = ((-1, 0), (1, 0), (0, -1), (0, 1))
HOUSE = "1"
EMPTY = "0"

n = int(input())
houses = [list(input().rstrip()) for _ in range(n)]

house_counts = []
bfs = deque() #너비 우선 탐색
for r in range(n):
    for c in range(n):
        if houses[r][c] == HOUSE: #집이 있음
            house_count = 1
            bfs.append((r, c))
            houses[r][c] = EMPTY
            while bfs: #인접한 집 확인
                start_r, start_c = bfs.popleft()
                for move_r, move_c in MOVE_DIR:
                    end_r = start_r + move_r
                    end_c = start_c + move_c
                    if end_r < 0 or end_r >= n or end_c < 0 or end_c >= n:
                        continue
                    if houses[end_r][end_c] == HOUSE: #인접한 집이 있음
                        house_count += 1
                        bfs.append((end_r, end_c))
                        houses[end_r][end_c] = EMPTY
            house_counts.append(house_count)

house_counts.sort() #오름차순 정렬
print(len(house_counts), *house_counts, sep="\n") #단지 수와 단지당 집 수

File: Baekjoon/Class3/34.2667 단지번호붙이기.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const MOVE_DIR = [[-1, 0], [1, 0], [0, -1], [0, 1]];
const HOUSE = "1";
const EMPTY = "0";

const n = Number(input[0]);

const houses = [];
for(let i = 1; i <= n; i++)
    houses.push(input[i].split(""));

const houseCounts = [];
for(let r = 0; r < n; r++)
    for(let c = 0; c < n; c++)
        if(houses[r][c] === HOUSE) { // 집이 있음
            let houseCount = 1;
            let bfsHead = 0; // 너비 우선 탐색
            const bfs = [[r, c]];
            houses[r][c] = EMPTY;
            while(bfsHead < bfs.length) { // 인접한 집 확인
                const [startR, startC] = bfs[bfsHead++];
                for(const [moveR, moveC] of MOVE_DIR) {
                    const endR = startR + moveR;
                    const endC = startC + moveC;
                    if(endR < 0 || endR >= n || endC < 0 || endC >= n)
                        continue;
                    if(houses[endR][endC] === HOUSE) { // 인접한 집이 있음
                        houseCount++;
                        bfs.push([endR, endC]);
                        houses[endR][endC] = EMPTY;
                    }
                }
            }
            houseCounts.push(houseCount);
        }

houseCounts.sort((a, b) => a - b); // 오름차순 정렬
console.log(`${houseCounts.length}\n${houseCounts.join("\n")}`); // 단지 수와 단지당 집 수
